### PR TITLE
[Woo POS] Extract cart logic to `CartViewModel`

### PR DESCRIPTION
**Pull request by iamgabrielma**
Part of https://github.com/woocommerce/woocommerce-ios/issues/12998

## Description
This is the step 2 of https://github.com/woocommerce/woocommerce-ios/issues/12998 , here we move Cart-related actions and its handling to a separate `CartViewModel`.

Its design follows the same pattern as with the product list, we'll observe changes in the model from the dashboard and use these to coordinate and communicate between different subviews. In general:

- When the cart is submitted to the next stage of the flow via `submitCart(_:)`, emits the cart items to its subscribers.
- When the cart is modified to the previous stage of the flow via `addMoreToCart()`, emits a signal to its subscribers.
- All cart-related logic (add/delete/clear/...) happens internally on the VM.

## Changes
- Created `CartViewModel`
- Moved cart-related logic to this new VM
- Added unit tests

## Testing
* Run the POS, and observe the product selector and cart interactions behave normally:
  * Adding/removing/"clear all" items to the cart works normally
  * Tapping on "checkout" submits the order for next stage (totals + payment screens)
  * Tapping "add more" brings us back to the product list (after re-loading products from the network) and we can continue updating the order as needed.

Simulator Screen Recording - iPad Pro (12 9-inch) (6th generation) - 2024-07-01 at 15 50 33

There's 3 items that I left outside of this PR, since are not totally related to the refactor itself, and can be tackled separately:

- Mocking `CartViewModel` for preview simplicity: https://github.com/woocommerce/woocommerce-ios/issues/13207
- Rethink how we set the ID for `CartItem`: https://github.com/woocommerce/woocommerce-ios/issues/13209
- Dashboard tests: https://github.com/woocommerce/woocommerce-ios/issues/13210

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by iamgabrielma** [comment]
Only one review is needed. 

cc: @staskus just looping you in for awareness, but feel free to review/drop any comment as well 🙇

**Comment by bozidarsevo** [comment]
I see few more // TODO: 12998, will there be more extraction PRs for the PointOfSaleDashboardViewModel?

**Review by bozidarsevo — APPROVED** [review]
Just that question regarding TODOs 👍

File: WooCommerce/Classes/POS/Presentation/CartView.swift [hunk 1]
```diff
@@ -102,9 +106,18 @@ private extension CartView {
...
+import Combine
 #Preview {
-    CartView(viewModel: PointOfSaleDashboardViewModel(itemProvider: POSItemProviderPreview(),
-                                                      cardPresentPaymentService: CardPresentPaymentPreviewService(),
-                                                      orderService: POSOrderPreviewService()))
+    // TODO:
+    // Simplify this by mocking `CartViewModel`
+    // https://github.com/woocommerce/woocommerce-ios/issues/13207
+    let orderStageSubject = PassthroughSubject<PointOfSaleDashboardViewModel.OrderStage, Never>()
+    let orderStagePublisher = orderStageSubject.eraseToAnyPublisher()
```
Comment (jaclync) [inline WooCommerce/Classes/POS/Presentation/CartView.swift:115]: In SwiftUI preview, we could also pass a constant value like passing `Just< PointOfSaleDashboardViewModel.OrderStage >(.building)` to `CartViewModel(orderStage:)`. We can potentially have two previews, one for each order stage case.

Would like to see what you think about this change, I will merge the PR without updating this 🙂

**Comment by jaclync** [comment]
> I see few more // TODO: 12998, will there be more extraction PRs for the PointOfSaleDashboardViewModel?

👋  I added those TODO comments, and yes they will be moved/extracted from `PointOfSaleDashboardViewModel` to a different view model in https://github.com/woocommerce/woocommerce-ios/pull/13211.

File: WooCommerce/WooCommerceTests/POS/ViewModels/CartViewModelTests.swift [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        XCTAssertTrue(sut.itemsInCart.isNotEmpty)
+    }
+
+    func test_removeItemFromCart() {
+        /* TODO:
+         https://github.com/woocommerce/woocommerce-ios/issues/13209
+         The unique UUID for CartItem is set on init, but CartItem is only internal to addItemToCart()
+         We need to extract this to a separate function and assure that ID's are correct,
+         otherwise the UUID's for testing won't match
+         */
```
Comment (jaclync) [inline WooCommerce/WooCommerceTests/POS/ViewModels/CartViewModelTests.swift:77]: For this test case, I wonder if we could first add some items by using the cart item from `itemsInCart`. I added a simple test case in https://github.com/woocommerce/woocommerce-ios/pull/13191/commits/e15efe2ec908f30a63675e8ff334624a6819d8b5, feel free to add other edge cases in the TODO issue.

**Review by jaclync — APPROVED** [review]
Thanks for updating the preview and tests! I added a test case and tested it for 2 transactions again, LGTM :shipit: I'll merge it shortly to unblock https://github.com/woocommerce/woocommerce-ios/pull/13211.

File: WooCommerce/Classes/POS/Presentation/CartView.swift (hunk 1 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/POS/Presentation/CartView.swift:115]: Thanks for merging, appreciated! 🙇 

> In SwiftUI preview, we could also pass a constant value like passing Just< PointOfSaleDashboardViewModel.OrderStage >(.building) to CartViewModel(orderStage:). We can potentially have two previews, one for each order stage case.

That sounds good, I've updated https://github.com/woocommerce/woocommerce-ios/issues/13207 for reference

File: WooCommerce/WooCommerceTests/POS/ViewModels/CartViewModelTests.swift (hunk 2 above)
Comment (iamgabrielma) [reply WooCommerce/WooCommerceTests/POS/ViewModels/CartViewModelTests.swift:77]: Yes, initially I wanted to test only by `id` to assure that we remove the exact "line item" if a cart has multiple of the same `productID`, so if we happen to change the implementation, we wouldn't get a false positive from tests if we happen to have just another item with the same name or property in the list. 

That said I think testing it like this and adding the followup is good enough for the moment (better than having no test case!). Thanks for tackling it 🙇

**Pull request merged**